experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,697 @@
+      insReady = !insValid || outsReady;
+      if (insValid && outsReady) {
+        outs = ins;
+      }
+    }
+  }
+
+private:
+  unsigned dataWidth;
+};
+
+class TEHBDataless {
+public:
+  TEHBDataless() = default;
+  void exec(bool isReset, bool insValid, bool outsReady, bool &insReady,
+            bool &outsValid) {
+    if (isReset) {
+      insReady = 1;
+    } else {
+      insReady = !insValid || outsReady;
+    }
+    outsValid = insValid;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Cocrete Execution Models
+//===----------------------------------------------------------------------===//
+class SinkModel : public OpExecutionModel<handshake::SinkOp> {
+public:
+  using OpExecutionModel<handshake::SinkOp>::OpExecutionModel;
+  SinkModel(handshake::SinkOp sinkOp)
+      : OpExecutionModel<handshake::SinkOp>(sinkOp), op(sinkOp) {}
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+    auto *ins = getState<ControlConsumerRW>(rws[{op.getOperand(), op}]);
+    ins->ready = true;
+  }
+
+private:
+  handshake::SinkOp op;
+};
+
+class MulIModel : public OpExecutionModel<handshake::MulIOp> {
+public:
+  using OpExecutionModel<handshake::MulIOp>::OpExecutionModel;
+  MulIModel(handshake::MulIOp muliOp)
+      : OpExecutionModel<handshake::MulIOp>(muliOp), op(muliOp),
+        bitwidth(cast<handshake::ChannelType>(muliOp.getLhs().getType())
+                     .getDataBitWidth()),
+        muliJoin(2), muliMul4Stage(bitwidth), muliDelayBuffer(3),
+        muliOEHB(bitwidth) {}
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+    auto *lhs = getState<ChannelConsumerRW>(rws[{op.getLhs(), op}]);
+    auto *rhs = getState<ChannelConsumerRW>(rws[{op.getRhs(), op}]);
+    auto *result = getState<ChannelProducerRW>(rws[{op.getResult(), op}]);
+    auto lhsData = llvm::any_cast<APInt>(lhs->data);
+    auto rhsData = llvm::any_cast<APInt>(rhs->data);
+
+    bool joinValid = false, buffValid = false, oehbReady = false;
+    Any oehbDataIn, oehbDataOut;
+    APInt resData;
+
+    std::vector<bool> insValid{lhs->valid, rhs->valid};
+    std::vector<std::reference_wrapper<bool>> insReady{lhs->ready, rhs->ready};
+
+    for (int i = 0; i < 10; ++i) {
+      muliJoin.exec(isReset, insValid, oehbReady, insReady, joinValid);
+      muliMul4Stage.exec(isReset, oehbReady, lhsData, rhsData, resData);
+      result->data = resData;
+      muliDelayBuffer.exec(isReset, joinValid, oehbReady, buffValid);
+      muliOEHB.exec(isReset, oehbDataIn, buffValid, result->ready, oehbReady,
+                    oehbDataOut, result->valid);
+    }
+  }
+
+private:
+  handshake::MulIOp op;
+  unsigned bitwidth;
+  Join muliJoin;
+  Mul4Stage muliMul4Stage;
+  DelayBuffer muliDelayBuffer;
+  OEHB muliOEHB;
+};
+
+class ReturnModel : public OpExecutionModel<handshake::ReturnOp> {
```

Comment:
You should probably just have an exec model for the TEHB and associate `handhsake::ReturnOp` operations to it directly, since a return is just a TEHB. FYI, in a few weeks I will probably delete the return operation, it is quite useless in the first place.

Reply:
Yeah, this difference between TEHB and Return is a mess.... However, some components (e g mux) use TEHB as internal component. I'm still thinking how to represent such cases better (maybe allow "empty operations" in exec models etc ).